Update repo: committing code updates for Lc->pKpi analysis

strogolo opened the pull request:
Dear @fcatalan92, @fgrosa,
I am doing this PR in order to push in the repository the updates that I have done for the non-prompt Lc analysis in pp collisions at 13 TeV.
In particular I have updated:
- GetRawYieldsDplusDs.py 
- ProjectDplusDsSparse.py
- ProjectDplusDsTree.py
- RunFullAnalysis.sh
and I have committed as new macro/files:
- ComputeEffAccWeightedAvg.py
- ComputeEffAccWeightedAvg.C
- config_LctopKpi_Fit_pp13TeV_basic.yml

Please let me know if something is not correct and has to be fixed.
Thanks and regards,
Stefano

fcatalan92 commented on ProjectDplusDsTree.py:
```diff
@@ -65,6 +67,12 @@
 bitFD = 3
 bitRefl = 4
+#define specific bits for Lc->pKpi resonant channel
+bitLcNonRes = 9
+bitLcLambda1520 = 10
+bitLcKStar = 11
+bitLcDelta = 12
```
```suggestion
bitsLcResonance = [9, 10, 11, 12] # ['NonRes', 'Lambda1520', 'KStar', 'Delta']
```

strogolo replied:
Done

fcatalan92 commented on ProjectDplusDsTree.py:
```diff
@@ -147,13 +155,29 @@
     if 'cand_type' in dataFramePrompt.columns: #if not filtered tree, select only prompt and not reflected
         dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitSignal, bitPrompt], 'and')
         dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitRefl], 'not')
+        if args.LctopKpireso == 1:
+            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitLcNonRes], 'and')
+        if args.LctopKpireso == 2:
+            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitLcKStar], 'and')
+        if args.LctopKpireso == 3:
+            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitLcDelta], 'and')
+        if args.LctopKpireso == 4:
+            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitLcLambda1520], 'and')
```
You can rewrite this by making a list of the different bits and accessing the list with `args.LctopKpireso` to avoid the copy-paste

```suggestion
        if args.LctopKpireso in range(1, 5):
            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitsLcResonance[args.LctopKpireso]], 'and')
```

strogolo replied:
Done

fcatalan92 commented on ProjectDplusDsTree.py:
```diff
@@ -147,13 +155,29 @@
     if 'cand_type' in dataFrameFD.columns: #if not filtered tree, select only FD and not reflected
         dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitSignal, bitFD], 'and')
         dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitRefl], 'not')
+        if args.LctopKpireso == 1:
+            dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitLcNonRes], 'and')
+        if args.LctopKpireso == 2:
+            dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitLcKStar], 'and')
+        if args.LctopKpireso == 3:
+            dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitLcDelta], 'and')
+        if args.LctopKpireso == 4:
+            dataFrameFD = FilterBitDf(dataFrameFD, 'cand_type', [bitLcLambda1520], 'and')
```
```suggestion
       if args.LctopKpireso in range(1, 5):
            dataFramePrompt = FilterBitDf(dataFramePrompt, 'cand_type', [bitsLcResonance[args.LctopKpireso]], and')
```

strogolo replied:
Done

fcatalan92 commented on RunFullAnalysis.sh:
```diff
@@ -175,17 +206,25 @@ if $DoMCProjection; then
-    elif [ "${MultWeightsFileName}" != "" ] && [ "${MultWeightsHistoName}" != "" ]; then
-        echo $(tput setaf 6) Using ${MultWeightsHistoName} mult weights from ${MultWeightsFileName} $(tput sgr0)
-        python3 ${ProjectScript} ${cfgFileMC} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml  ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root --multweights ${MultWeightsFileName} ${MultWeightsHistoName}
+    if [ ${Particle} == "LctopKpi" ]; then
+      python3 ${ProjectScript} ${cfgFileMC} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root
+      python3 ${ProjectScript} ${cfgFileMCNonRes} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml  ${OutDirEfficiency}/Distr_${Particle}_NonRes_MC${CutSets[$iCutSet]}.root --LctopKpireso 1
+      python3 ${ProjectScript} ${cfgFileMCKStar} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml  ${OutDirEfficiency}/Distr_${Particle}_KStar_MC${CutSets[$iCutSet]}.root --LctopKpireso 2
+      python3 ${ProjectScript} ${cfgFileMCDelta} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml  ${OutDirEfficiency}/Distr_${Particle}_Delta_MC${CutSets[$iCutSet]}.root --LctopKpireso 3
+      python3 ${ProjectScript} ${cfgFileMCLambda1520} ${CutSetsDir}/cutset${CutSets[$iCutSet]}.yml  ${OutDirEfficiency}/Distr_${Particle}_Lambda1520_MC${CutSets[$iCutSet]}.root --LctopKpireso 4
```
Here you are removing the possibility to apply pT weights and other corrections for the Lc->pkpi. Maybe we need to think about how to improve the approach and build the command incrementally

strogolo replied:
I will implement the possiblity to apply pT weights when I'll have to do it.

fcatalan92 commented on RunFullAnalysis.sh:
```diff
@@ -218,17 +257,53 @@ fi
   do
-    echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root $(tput sgr0)
-    python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}${CutSets[$iCutSet]}.root --batch
+    if [ ${Particle} == "LctopKpi" ]; then
+      echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root $(tput sgr0)
+      python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}${CutSets[$iCutSet]}.root --batch
+      echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_NonRes_MC${CutSets[$iCutSet]}.root $(tput sgr0)
+      python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_NonRes_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}_NonRes${CutSets[$iCutSet]}.root --batch
+      echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_KStar_MC${CutSets[$iCutSet]}.root $(tput sgr0)
+      python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_KStar_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}_KStar${CutSets[$iCutSet]}.root --batch
+      echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_Delta_MC${CutSets[$iCutSet]}.root $(tput sgr0)
+      python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_Delta_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}_Delta${CutSets[$iCutSet]}.root --batch
+      echo $(tput setaf 4) Compute efficiency from ${OutDirEfficiency}/Distr_${Particle}_Lambda1520_MC${CutSets[$iCutSet]}.root $(tput sgr0)
+      python3 ComputeEfficiencyDplusDs.py ${cfgFileFit} ${Cent} ${OutDirEfficiency}/Distr_${Particle}_Lambda1520_MC${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Efficiency_${Particle}_Lambda1520${CutSets[$iCutSet]}.root --batch
```
here a for loop could be used

strogolo replied:
Done

fcatalan92 commented on RunFullAnalysis.sh:
```diff
@@ -218,17 +257,53 @@ fi
-    python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}.root --batch
+    if [ ${Particle} == "LctopKpi" ]; then
+      echo $(tput setaf 4) Compute efficiency times acceptance $(tput sgr0)
+      python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}.root --batch
+      python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}_NonRes${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}_NonRes${CutSets[$iCutSet]}.root --batch
+      python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}_KStar${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}_KStar${CutSets[$iCutSet]}.root --batch
+      python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}_Delta${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Delta${CutSets[$iCutSet]}.root --batch
+      python3 CombineAccTimesEff.py ${OutDirEfficiency}/Efficiency_${Particle}_Lambda1520${CutSets[$iCutSet]}.root ${accFileName} ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Lambda1520${CutSets[$iCutSet]}.root --batch
```
maybe for loop also here

strogolo replied:
Done

fcatalan92 reviewed (changes requested):
Hi @strogolo, thank you for your contribution. I added a few comments.
Moreover, I would prefer to have only one macro for the weighted average of the efficiency. I would keep `ComputeEffAccWeightedAvg.py`, we have still a few macros that are both in C++ and in Python but it is not a good idea.

fcatalan92 commented on RunFullAnalysis.sh:
```diff
@@ -218,17 +257,53 @@ fi
+if $DoAccEffRw; then
+  for (( iCutSet=0; iCutSet<${arraylength}; iCutSet++ ));
+  do
+    if [ ${Particle} == "LctopKpi" ]; then
+      echo $(tput setaf 4) Compute re-weighted efficiency times acceptance $(tput sgr0)
+      python3 ComputeEffAccWeightedAvg.py ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_NonRes${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_KStar${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Delta${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Lambda15 ...
```
For loop also here

strogolo replied:
No, here the loop cannot be used. It is just one single command line.

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,72 @@
+parser.add_argument("--batch", help="suppress video output", action="store_true")
+args = parser.parse_args()
+
+gROOT.SetBatch(args.batch)
+SetGlobalStyle(padleftmargin=0.14, padbottommargin=0.12, titlesize=0.045, labelsize=0.04)
```
In the previous version of the script I was not drawing the histograms, while now I'm doing it in the latest update

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,72 @@
+hEffLambda1520Prompt = effFileLambda1520.Get('hAccEffPrompt')
+hEffLambda1520FD = effFileLambda1520.Get('hAccEffFD')
+SetObjectStyle(hEffLambda1520Prompt, color=kRed+1, markerstyle=kFullCircle)
+SetObjectStyle(hEffLambda1520FD, color=kAzure+4, markerstyle=kOpenSquare, markersize=1.5, linewidh=2, linestyle=7)
```
I re-wrote everything using loops in python. It took me sometime to get everything working since I am not super-expert of python. Nevertheless, I checked the output and the result is identical to the old version of the script.

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,72 @@
+hEffAllPrompt = effFileAll.Get('hAccEffPrompt')
+hEffAllFD = effFileAll.Get('hAccEffFD')
+SetObjectStyle(hEffAllPrompt, color=kRed+1, markerstyle=kFullCircle)
+SetObjectStyle(hEffAllFD, color=kAzure+4, markerstyle=kOpenSquare, markersize=1.5, linewidh=2, linestyle=7)
```
It is needed in the last version of the script in order to draw the histograms.

strogolo commented on GetRawYieldsDplusDs.C:
```diff
@@ -28,7 +28,7 @@
 
 enum {k010, k3050, k6080, kpp5TeVPrompt, kpp5TeVFD, kpp13TeVPrompt, kpp13TeVFD};
-enum {kDplus, kD0, kDs, kLc, kDstar};
+enum {kDplus, kD0, kDs, kLctopKpi, kLctopK0s, kDstar};
```
The `kLctopK0s` case was already present in the `enum`, now I fixed the discrepancies in the macro.

strogolo commented on GetRawYieldsDplusDs.C:
```diff
@@ -68,8 +68,10 @@ int GetRawYieldsDplusDs(int cent, bool isMC, TString infilename, TString refFile
         particle=kD0;
-    } else if (ParticleName=="Lc"){
-        particle=kLc;
+    } else if (ParticleName=="LctopKpi"){
+        particle=kLctopKpi;
+    } else if (ParticleName=="LctopK0s"){
+        particle=kLctopK0s;
```
Done

strogolo commented on RunFullAnalysis.sh:
```diff
@@ -218,17 +257,53 @@ fi
+      echo $(tput setaf 4) Compute re-weighted efficiency times acceptance $(tput sgr0)
+      python3 ComputeEffAccWeightedAvg.py ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_NonRes${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_KStar${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Delta${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Lambda15 ...
+      #echo '.x ComputeEffAccWeightedAvg.C+('${OutDirEfficiency}','${Particle}','${CutSets[$iCutSet]}' ,"Eff_times_Acc_'${Particle}${CutSets[$iCutSet]}'_rw.root")' | root -l -b
+      #echo '.q'
```
Done

strogolo commented:
> Hi @strogolo, thank you for your contribution. I added a few comments. Moreover, I would prefer to have only one macro for the weighted average of the efficiency. I would keep `ComputeEffAccWeightedAvg.py`, we have still a few macros that are both in C++ and in Python but it is not a good idea.

Hi @fcatalan92, I have implemented almost all of your suggestions. Regarding the double version of the macro ComputeEffAccWeightedAvg (in `python` and in `C++`) I understand your point. However, I just want to inform you that on my laptop I'm using the C++ version, while the python version was written for completeness for your repository. Thus, I was just wondering if it is good to have at least a version of the macro that I'm currently using as documentation. 

Thanks and regards,
Stefano

fcatalan92 commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+#effC, wC = 0,0
+nPtBins = hEffPrompt[0].GetNbinsX()
+for iPt in range(nPtBins):
+    print('==PROMPT==')
+    print(iPt)
```
```suggestion
```

strogolo replied:
Done

fcatalan92 commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+    for histo, br, in zip(hEffPrompt[1:],BR[1:]):
+        effC += histo.GetBinContent(iPt+1)*br
+        wC += br
+        print('--> B.R.: %f' % (br))
+        print(effC)
+        print(wC)
```
```suggestion
```

strogolo replied:
Done

fcatalan92 commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+    hEffCw.SetBinContent(iPt+1, (effC/wC))
+    hEffCw.SetBinError(iPt+1, hEffPrompt[0].GetBinError(iPt+1))
+
+    print('==FD==')
+    print(iPt)
```
```suggestion
```

strogolo replied:
Done

fcatalan92 commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+    for histo, br, in zip(hEffFD[1:],BR[1:]):
+        effB += histo.GetBinContent(iPt+1)*br
+        wB += br
+        print('--> B.R.: %f' % (br))
+        print(effB)
+        print(wB)
```
```suggestion
```

strogolo replied:
Done

fcatalan92 commented:
Hi @strogolo, thank you for implementing the changes. I have just a few suggestions more.

On the `ComputeEffAccWeightedAvg`, have you checked that the python version is equivalent to the C++ one (same outputs)? If yes, I think we can merge only the former one. @fgrosa What do you think?

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+hEffBw = hEffFD[0].Clone("hAccEffFD")
+
+BR = [6.28 * 1e-02, 3.5 * 1e-02, 1.96 * 0.667 * 1e-02, 1.08 * 1e-02, 2.2 * 0.225 * 1e-02]
+#effC, wC = 0,0
```
Fixed

fcatalan92 replied:
This should be removed not decommented

strogolo replied:
Don't we need to define `effC, wC`before using them?

strogolo replied:
@fcatalan92 if you can suggest the change, I'll accept it and merge it immediately

strogolo replied:
Done

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,100 @@
+hEffPrompt, hEffFD = [], []
+
+for iReso, fileName, in enumerate(inputFile):
+    print(iReso)
```
Done

strogolo commented:
> Hi @strogolo, thank you for implementing the changes. I have just a few suggestions more.
> 
> On the `ComputeEffAccWeightedAvg`, have you checked that the python version is equivalent to the C++ one (same outputs)? If yes, I think we can merge only the former one. @fgrosa What do you think?

Hi @fcatalan92, I have removed the debug print as requested. 
Concerning your question: I did not check in details that the output of the python version is equivalent to the C++ one. From a quick look it seems everything equivalent, but I did not perform an apple-to-apple comparison because the main macro I'm using for the analysis is the C++ one, while I wrote the python for completeness and because someone could prefer to use the python version. Look, if it is such a huge problem I can remove `ComputeEffAccWeightedAvg` (both python and C++) since it is a macro specific for the `Lc->pKpi` analysis and just keep it locally updated.

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+        effBw->Reset("icse");
+
+
+        for(int k=0; k<effC[0]->GetNbinsX(); k++){
+            effCw->SetBinContent(k + 1, ((effC[0]->GetBinContent(k + 1) * BR[0]) + (effC[1]->GetBinContent(k + 1) * BR[1]) + (effC[2]->GetBinContent(k + 1) * BR[2]) + (effC[3]->GetBinContent(k + 1) * BR[3])) / (BR[0] + BR[1] + BR[2] + BR[3]));
+            effCw->SetBinError(k + 1, effC[4]->GetBinError(k + 1));
+
+            effBw->SetBinContent(k + 1, ((effB[0]->GetBinContent(k + 1) * BR[0]) + (effB[1]->GetBinContent(k + 1) * BR[1]) + (effB[2]->GetBinContent(k + 1) * BR[2]) + (effB[3]->GetBinContent(k + 1) * BR[3])) / (BR[0] + BR[1] + BR[2] + BR[3]));
+            effBw->SetBinError(k + 1, effB[4]->GetBinError(k + 1));
+        }
```
```suggestion
        for(int iPt=0; iPt<effC[0]->GetNbinsX(); iPt++) {
            double weightEffC = 0., weightUncEffC = 0., weightEffB = 0., weightUncEffB = 0., sumOfBR = 0.;
            for(int iCh=0; iCh<4; iCh++) {
                weightEffC += effC[iCh]->GetBinContent(iPt + 1) * BR[iCh];
                weightUncEffC += effC[iCh]->GetBinError(iPt + 1) * effC[iCh]->GetBinError(iPt + 1) * BR[iCh] * BR[iCh];
                weightEffB += effB[iCh]->GetBinContent(iPt + 1) * BR[iCh];
                weightUncEffB += effB[iCh]->GetBinError(iPt + 1) * effB[iCh]->GetBinError(iPt + 1) * BR[iCh] * BR[iCh];
                sumOfBR += BR[iCh];
            }
            weightEffC /= sumOfBR;
            weightUncEffC = TMath::Sqrt(weightUncEffC) / sumOfBR;
            weightEffB /= sumOfBR;
            weightUncEffB = TMath::Sqrt(weightUncEffB) / sumOfBR;
            effC[iCh]->GetBinContent(iPt + 1, weightEffC);
            effC[iCh]->GetBinError(iPt + 1, weightUncEffC);
            effB[iCh]->GetBinContent(iPt + 1, weightEffB);
            effB[iCh]->GetBinError(iPt + 1, weightUncEffB);
        }
```

strogolo replied:
@fgrosa thanks for this change. Just for clarification: the way the uncertainty were taken in the first version was exactly the same used in the previous Lc analyses. Nevertheless I agree with your change which makes the uncertainty calculation more correct.

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+#include "TFile.h"
+#include "TGaxis.h"
+#include "TDatabasePDG.h"
+
```
```suggestion
#include "TMath.h"
```

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+            Form("%s/Eff_times_Acc_%s_KStar%s.root", effdir.Data(), particle.Data(), cutset.Data()), //KStar
+            Form("%s/Eff_times_Acc_%s_Delta%s.root", effdir.Data(), particle.Data(), cutset.Data()), //Delta
+            Form("%s/Eff_times_Acc_%s_Lambda1520%s.root", effdir.Data(), particle.Data(), cutset.Data()), //Lambda1520
+            Form("%s/Eff_times_Acc_%s%s.root", effdir.Data(), particle.Data(), cutset.Data()) //All
```
```suggestion
```

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+        };
+        
+        TFile *file[5];
+        for (int i = 0; i < 5; i++){
```
```suggestion
        for (int i = 0; i < 4; i++){
```

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+            Form("%s/Eff_times_Acc_%s%s.root", effdir.Data(), particle.Data(), cutset.Data()) //All
+        };
+        
+        TFile *file[5];
```
```suggestion
        TFile *file[4];
```

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+            2.2 * 0.225 * 1e-02 //L1520
+        };
+
+        TH1D *effC[5];
+        TH1D *effB[5];
```
```suggestion
        TH1D *effC[4];
        TH1D *effB[4];
```

fgrosa commented on ComputeEffAccWeightedAvg.C:
```diff
@@ -0,0 +1,82 @@
+
+        TH1D *effC[5];
+        TH1D *effB[5];
+        for(int i = 0; i < 5; i++){
```
```suggestion
        for(int i = 0; i < 4; i++){
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+effC, wC = 0,0
+nPtBins = hEffPrompt[0].GetNbinsX()
+for iPt in range(nPtBins):
+    effC, wC = 0,0
+    for histo, br, in zip(hEffPrompt[1:],BR[1:]):
+        effC += histo.GetBinContent(iPt+1)*br
+        wC += br
+    hEffCw.SetBinContent(iPt+1, (effC/wC))
+    hEffCw.SetBinError(iPt+1, hEffPrompt[0].GetBinError(iPt+1))
+
+    effB, wB = 0,0
+    for histo, br, in zip(hEffFD[1:],BR[1:]):
+        effB += histo.GetBinContent(iPt+1)*br
+        wB += br
+    hEffBw.SetBinContent(iPt+1, (effB/wB))
+    hEffBw.SetBinError(iPt+1, hEffFD[0].GetBinError(iPt+1))
```
```suggestion
    effC, effB, uncEffC, uncEffB, sumOfW = (0. for _ in range(5))
    for histo, br, in zip(hEffPrompt, BR):
        effC += histo.GetBinContent(iPt+1) * br
        effB += histo.GetBinContent(iPt+1) * br
        uncEffC += histo.GetBinError(iPt+1)**2 * br**2
        uncEffB += histo.GetBinError(iPt+1)**2 * br**2
        sumOfW += br
    hEffCw.SetBinContent(iPt+1, (effC/sumOfW))
    hEffCw.SetBinError(iPt+1, np.sqrt(uncEffC)/sumOfW)
    hEffBw.SetBinContent(iPt+1, (effB/sumOfW))
    hEffBw.SetBinError(iPt+1, np.sqrt(uncEffB)/sumOfW)
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+from utils.StyleFormatter import SetGlobalStyle, SetObjectStyle
+
+parser = argparse.ArgumentParser(description='Arguments')
+parser.add_argument('effFileAll', metavar='text', default='')
```
```suggestion
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+gROOT.SetBatch(args.batch)
+SetGlobalStyle(padleftmargin=0.14, padbottommargin=0.12, titlesize=0.045, labelsize=0.04)
+
+inputFile = [args.effFileAll, args.effFileNonRes, args.effFileKStar, args.effFileDelta, args.effFileLambda1520]
```
```suggestion
inputFile = [args.effFileNonRes, args.effFileKStar, args.effFileDelta, args.effFileLambda1520]
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+    hEffPrompt[iReso].SetDirectory(0)
+    hEffFD[iReso].SetDirectory(0)
+    SetObjectStyle(hEffPrompt[iReso], color=colorMarkerPrompt[iReso], markerstyle=styleMarkerPrompt[iReso])
+    SetObjectStyle(hEffFD[iReso], color=colorMarkerFD[iReso], markerstyle=styleMarkerFD[iReso], markersize=1.5, linewidh=2, linestyle=7)
```
```suggestion
    SetObjectStyle(
        hEffFD[iReso],
        color=colorMarkerFD[iReso],
        markerstyle=styleMarkerFD[iReso],
        markersize=1.5, linewidh=2, linestyle=7
    )
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+    hEffFD.append(infile.Get('hAccEffFD'))
+    hEffPrompt[iReso].SetDirectory(0)
+    hEffFD[iReso].SetDirectory(0)
+    SetObjectStyle(hEffPrompt[iReso], color=colorMarkerPrompt[iReso], markerstyle=styleMarkerPrompt[iReso])
```
```suggestion
    SetObjectStyle(
        hEffPrompt[iReso],
        color=colorMarkerPrompt[iReso],
        markerstyle=styleMarkerPrompt[iReso]
    )
```

fgrosa commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,89 @@
+hEffCw = hEffPrompt[0].Clone("hAccEffPrompt")
+hEffBw = hEffFD[0].Clone("hAccEffFD")
+
+BR = [6.28 * 1e-02, 3.5 * 1e-02, 1.96 * 0.667 * 1e-02, 1.08 * 1e-02, 2.2 * 0.225 * 1e-02]
```
```suggestion
BR = [3.5 * 1e-02, 1.96 * 0.667 * 1e-02, 1.08 * 1e-02, 2.2 * 0.225 * 1e-02]
```

fgrosa commented on RunFullAnalysis.sh:
```diff
@@ -218,17 +258,41 @@ fi
+  do
+    if [ ${Particle} == "LctopKpi" ]; then
+      echo $(tput setaf 4) Compute re-weighted efficiency times acceptance $(tput sgr0)
+      python3 ComputeEffAccWeightedAvg.py ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_NonRes${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_KStar${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Delta${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Lambda15 ...
```
```suggestion
      python3 ComputeEffAccWeightedAvg.py ${OutDirEfficiency}/Eff_times_Acc_${Particle}_NonRes${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_KStar${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Delta${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}_Lambda1520${CutSets[$iCutSet]}.root ${OutDirEfficiency}/Eff_times_Acc_${Particle}${CutSets[$iCutSet]}_rw.root
```

fgrosa reviewed (commented):
Ciao @strogolo thanks a lot! I have a couple of comments that just go in the direction to simplify a bit the code (if they are fine with you you can just click on "commit suggestion"). For the pT weights, indeed it will have to be done, but it's ok for me to do it in a second step (also because in the meantime we might want to merge this PR https://github.com/DmesonAnalysers/DmesonAnalysis/pull/249). For the c++/python script I have no strong opinion, it is fine for me to keep only one of the two or both.

fcatalan92 commented:
@strogolo Thank you for the updates. I made just a minor comment.
IMHO, the most important thing is to have only one macro to maintain. So, if it is not possible to validate the python one (my preferred option), I would keep only the C++ macro.

strogolo commented on ComputeEffAccWeightedAvg.py:
```diff
@@ -0,0 +1,95 @@
+hEffBw = hEffFD[0].Clone("hAccEffFD")
+
+BR = [3.5 * 1e-02, 1.96 * 0.667 * 1e-02, 1.08 * 1e-02, 2.2 * 0.225 * 1e-02]
+effC, wC = 0,0
```
```suggestion

```
